src/charm.py

```diff
@@ -0,0 +1,553 @@
+                ),
+            ],
+        )
+
+    def _write_bessd_config_file(
+        self,
+        upf_hostname: str,
+        upf_mode: str,
+        access_interface_name: str,
+        core_interface_name: str,
+        dnn: str,
+        pod_share_path: str,
+    ) -> None:
+        """Write the configuration file for the 5G UPF service."""
+        jinja2_environment = Environment(loader=FileSystemLoader("src/templates/"))
+        template = jinja2_environment.get_template(f"{CONFIG_FILE_NAME}.j2")
+        content = template.render(
+            upf_hostname=upf_hostname,
+            mode=upf_mode,
+            access_interface_name=access_interface_name,
+            core_interface_name=core_interface_name,
+            dnn=dnn,
+            pod_share_path=pod_share_path,
+        )
+        self._bessd_container.push(
+            path=f"{BESSD_CONTAINER_CONFIG_PATH}/{CONFIG_FILE_NAME}", source=content
+        )
+        logger.info(f"Pushed {CONFIG_FILE_NAME} config file")
+
+    @property
+    def _upf_hostname(self) -> str:
+        """Returns the UPF hostname.
+
+        Returns:
+            str: UPF Hostname
+        """
+        return f"{self.model.app.name}.{self.model.name}.svc.cluster.local"
+
+    @property
+    def _bessd_config_file_is_written(self) -> bool:
+        """Returns whether the bessd config file was written to the workload container.
+
+        Returns:
+            bool: Whether the bessd config file was written
+        """
+        if not self._bessd_container.exists(f"{BESSD_CONTAINER_CONFIG_PATH}/{CONFIG_FILE_NAME}"):
+            logger.info(f"Config file is not written: {CONFIG_FILE_NAME}")
+            return False
+        logger.info("Config file is written")
+        return True
+
+    @property
+    def _pfcp_agent_config_file_is_written(self) -> bool:
+        """Returns whether the pfcp agent config file was written to the workload container.
+
+        Returns:
+            bool: Whether the pfcp agent config file was written
+        """
+        if not self._pfcp_agent_container.exists(
+            f"{PFCP_AGENT_CONTAINER_CONFIG_PATH}/{CONFIG_FILE_NAME}"
+        ):
+            logger.info(f"Config file is not written: {CONFIG_FILE_NAME}")
+            return False
+        logger.info("Config file is written")
+        return True
+
+    def _configure_bessd_workload(self, event: EventBase) -> None:
+        """Handles events for configuring bessd container.
+
+        Args:
+            event: EventBase
+        """
+        invalid_configs = self._get_invalid_configs()
+        if invalid_configs:
+            self.unit.status = BlockedStatus(
+                f"The following configurations are not valid: {invalid_configs}"
+            )
+            event.defer()
+            return
+        if not self._bessd_container.can_connect():
+            self.unit.status = WaitingStatus("Waiting to be able to connect to the container")
+            event.defer()
```

Comment:
I think we can remove this `event.defer()` as pebble will trigger. Happens the same in the `smf` as noted by @ghislainbourgeois.

Reply:
Good catch!